rhino/src/main/java/org/mozilla/javascript/NativeString.java

```diff
@@ -772,6 +782,80 @@ else if (Normalizer.Form.NFC.name().equals(formStr))
 
                         return str.substring(k, k + 1);
                     }
+                case Id_isWellFormed:
+                    {
+                        CharSequence str =
+                                ScriptRuntime.toCharSequence(
+                                        requireObjectCoercible(cx, thisObj, f));
+                        int len = str.length();
+                        boolean foundLeadingSurrogate = false;
+                        for (int i = 0; i < len; i++) {
+                            char c = str.charAt(i);
+                            if (NativeJSON.isLeadingSurrogate(c)) {
+                                if (foundLeadingSurrogate) {
+                                    return false;
+                                }
+                                foundLeadingSurrogate = true;
+                            } else if (NativeJSON.isTrailingSurrogate(c)) {
+                                if (!foundLeadingSurrogate) {
+                                    return false;
+                                }
+                                foundLeadingSurrogate = false;
+                            } else if (foundLeadingSurrogate) {
+                                return false;
+                            }
+                        }
+                        return !foundLeadingSurrogate;
+                    }
+                case Id_toWellFormed:
+                    {
+                        CharSequence str =
+                                ScriptRuntime.toCharSequence(
+                                        requireObjectCoercible(cx, thisObj, f));
+                        // true represents a surrogate pair
+                        // false represents a singular surrogate
+                        // normal characters aren't present
+                        Map<Integer, Boolean> surrogates = new HashMap<>();
+
+                        int len = str.length();
+                        char prev = 0;
+                        int firstSurrogateIndex = -1;
+                        for (int i = 0; i < len; i++) {
+                            char c = str.charAt(i);
+
+                            if (NativeJSON.isLeadingSurrogate(prev)
+                                    && NativeJSON.isTrailingSurrogate(c)) {
+                                surrogates.put(Integer.valueOf(i - 1), Boolean.TRUE);
```

Comment:
I used the Boxed types for the map because I saw a commit from a few years ago that reduced the amount of auto boxing.